File: Python/55-链表中环的入口结点.py
# -*- coding:utf-8 -*-
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def EntryNodeOfLoop(self, pHead):
        # write code here
        if not pHead or not pHead.next or not pHead.next.next:
            return None
        slow = pHead.next
        fast = pHead.next.next
        while slow != fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        if slow == fast:
            slow = pHead
            while slow != fast:
                slow = slow.next
                fast = fast.next
            return slow
        return None

File: C++/55-链表中环的入口结点.cpp
/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
        val(x), next(NULL) {
    }
};
*/
class Solution
{
public:
    ListNode *EntryNodeOfLoop(ListNode *pHead)
    {
        if (!pHead || !pHead->next || !pHead->next->next)
            return nullptr;
        ListNode *slow = pHead->next, *fast = pHead->next->next;
        while (slow != fast && fast->next)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        if (slow == fast)
        {
            slow = pHead;
            while (slow != fast)
            {
                slow = slow->next;
                fast = fast->next;
            }
            return slow;
        }
        return nullptr;
    }
};